### ISSUE TITLE
Generic Extensions (Actions) Proposal

### ISSUE THREAD

**Issue by avade** [openservicebrokerapi/servicebroker#114]
Following the discussion in https://github.com/openservicebrokerapi/servicebroker/issues/107 this issue aims to create a proposal for "Service Broker Actions".

---

## Problem 

*As a service broker author, I want to be able to optionally offer operations beyond "provision, update, bind, unbind, delete" so that I can provide new features to users of my service via the service broker API*

## Examples

- Backup and restore functionality (previous discussion https://github.com/openservicebrokerapi/servicebroker/issues/107)
- Schedule things via the service broker

## Workflows

Service clients can:

- Discover the actions available from a service broker
- Trigger a specific action synchronously
- Trigger a specific action asynchronously
- Pass input parameters to the action
- Discover and validate input parameters (input parameter schema)

Service authors can:

- Optionally define actions at service broker registration time

## History

https://github.com/openservicebrokerapi/servicebroker/issues/107 
https://docs.google.com/document/d/1Y5vwWjvaUIwHI76XU63cAS8xEOJvN69-cNoCQRqLPqU/edit#heading=h.kzfeyua6md59

## Next steps

1. [Proposal document]( https://docs.google.com/document/d/1Dhs4RrvXxyDcmK8LdxbkHcEb_HrpiA4XqVwaIy2sTXg)
1. Review on OSBAPI WG call

**Comment by gberche-orange** [comment]
Here is a [google docs placeholder](https://docs.google.com/document/d/1Dhs4RrvXxyDcmK8LdxbkHcEb_HrpiA4XqVwaIy2sTXg/edit). I'll try fleshing some ideas there in the coming days. Anyone feel free to start earlier if they have some bandwidth.

@avade do you have pointers to the WG calendar and dialin/minutes details ?

**Comment by avade** [comment]
@gberche-orange Perfect thanks! - @shalako can add you to the invite.

**Comment by shalako** [comment]
Sent. Guillaume, consider that dial-in info will change for next week's call. I'll be sending an update to the invite shortly.

**Comment by gberche-orange** [comment]
thanks @shalako 

I've proposed a first draft into https://docs.google.com/document/d/1Dhs4RrvXxyDcmK8LdxbkHcEb_HrpiA4XqVwaIy2sTXg/edit 

I've moved the sharing mode to comment mode so that we can systematically get notifications from suggestions. Let me know the list of people to share with edit role as to accept/reject suggestions collaboratively.

**Comment by gberche-orange** [comment]
@avade any feedback on the proposal prior to the WG call today ?

**Comment by avade** [comment]
I will be taking a look this morning.

**Comment by bmelville** [comment]
Is there any update on this proposal? It's something we're interested in seeing as well, so would love to see it gain traction.

B.

**Comment by avade** [comment]
Right now I am focusing on the schema / removing CFisms.

I should get a chance to get stuck into this later next week, but I this
feature seems to unlock lots of interesting broker uses cases for Pivotal.

**Comment by jmcarp** [comment]
I just filed #148 to ask about this more generally, but I was wondering whether this proposal might allow broker-defined actions to be executed on pending service instances.

**Comment by gberche-orange** [comment]
@jmcarp the proposal doc mentions currency control for broker actions with different approaches. Broker actions are certainly a subresource of a service instance, but we had not yet detailed how/whether consistency would be required between the two.

For instance, if a backup action is on going, can the parent service instance be updated or deleted, or similarly to service bindings should all broker actions complete before the parent service instance can be deleted. 

Having more details on your use cases for broker actions would help refine the proposal, including during long-lasting service instance create/update operations.

**Comment by dbdd4us** [comment]
LGTM
Any Update?
Thanks

**Comment by duglin** [comment]
general direction looks good.  Couple of things
- I'd like to discuss this at the f2f in a few weeks
- in prep for that, could we get a version of the spec with the proposed edits?
- we'll probably want to include the ability to describe the parameters to the actions via schema
- we may want to consider defining an "Actions Profile" to define some common action names so that we don't get "backup" vs "back-up" - if that's a problem, not 100% sure yet.

**Comment by petereberlein** [comment]
The proposal document actually contains two options: actions as parameters in a PATCH request and actions modelled as resources. I would strongly prefer the latter one as it is more transparent and also better extensible for long running actions (delete them, get a status update, etc.)

**Comment by avade** [comment]
I would like to revisit the conversation about supporting B&R as a first class API concept. It is still Pivotals main use case for actions.

We can still progress with actions as a concept, but I would prioritise solving the B&R case first.

**Comment by petereberlein** [comment]
Addressing a concrete use case (like B&R) will help detailing the concept, but it should be generic enough to add further actions later on without the need of extending/changing this concept.

So maybe we can just use B&R as a concrete example for actions in general?

**Comment by avade** [comment]
The approach I was thinking was to put both solutions (actions and making B&R first class in the API) then we can compare and contrast.

**Comment by duglin** [comment]
We'd prefer to not overload the current APIs so the 2nd approach (dedicated resources) feels cleaner.  One other bit of feedback I got was that the actions should not be linked to the plans - we'd prefer if there wasn't a tight coupling between plans and actions. Also, since the list of available actions may vary based on the state of the instance it would be nicer to support `GET /v2/service_instances/:instance_id/actions` instead.

Are people comfortable with making a statement about the type of actions that should be included here? Meaning, they should be limited to admin type of actions that act upon the service instance. I don't think we want people to take actions that should be invoked by the end-user via the connection info in the binding and move them to be actions performed by the platform on their behalf.  I view this as the diff between action "on" the service vs actions "with" service service.

**Comment by mattmcneeney** [comment]
In the CF world, many of these actions are likely to be long-running, so we would like to support asynchronous calls, with the ability to fetch the current status of any actions and cancelling a pending action. I'd also like to see a branch (or multiple) with the proposed changes so we can discuss the proposed implementation.

**Comment by gberche-orange** [comment]
I started detailing how a backup use-case could like with the 2nd approach. I called them "snapshots" to reflect that they are triggered by application developers at the given point in time (as to avoid confusion with service operators scheduled backups intended for data loss/durability)

In terms of UX, the end users would see the following list of actions for a given data service instance (say a RDBMS service) inferred from the broker catalog response, along with associated input parameters and response format (ommited here)
* create-snapshot
* restore-snapshot
* list-snapshots
* delete-snaphot

A possible REST calls flow would then be:

```
Requesting a snapshot (async)
curl -X PUT /v2/service_instances/:instance_id/actions_instances/1 -d "actions=create-snapshot&parameters={name=snapshot_A}"
202 ACCEPTED

Checking snapshot completion: in progress
curl -X GET /v2/service_instances/:instance_id/actions_instances/1
200 OK
state="in progress"
description="creating snapshot  10%"

Checking snaphot completion: completed
curl -X GET /v2/service_instances/:instance_id/actions_instances/1
200 OK
state="suceeded"
description="snapshot  id 1 created successfully."
resource={name=snapshot_A, size=100MB, download-url=...}

Requesting a new snapshot:

curl -X PUT /v2/service_instances/:instance_id/actions_instances/2 -d "actions=create-snapshot&parameters={name=snapshot_B}"
202 ACCEPTED

Checking snaphot completion: in progress (stalled)
curl -X GET /v2/service_instances/:instance_id/actions_instances/2
200 OK
state="in progress"
description="creating snapshot  10%"

Cancelling a pending snaphot operation:
curl -X DELETE /v2/service_instances/:instance_id/actions_instances/2
200 OK

Requesting lists of snapshots (synchronous)

curl -X PUT /v2/service_instances/:instance_id/actions_instances/3 -d "action=list-snapshots"
201 CREATED

curl -X GET /v2/service_instances/:instance_id/actions_instances/3
state="suceeded"
description="list of snapshots successfully completed."
resource=[{name=snapshot_A, size=100MB, download-url=...} , ]

Restore a given snapshot in time:

curl -X PUT /v2/service_instances/:instance_id/actions_instances/4 -d "action=restore-snapshot&parameters={name=snapshot_A}"
202 ACCEPTED

Checking restore completion: in progress
curl -X GET /v2/service_instances/:instance_id/actions_instances/4
200 OK
state="in progress"
description="restoring snapshot  10%"

Checking restore completion: completed
curl -X GET /v2/service_instances/:instance_id/actions_instances/4
200 OK
state="suceeded"
description="snapshot  id 1 created successfully."
resource={size=100MB,download-url=...}

Delete a given snapshot
curl -X PUT /v2/service_instances/:instance_id/actions_instances/4 -d "action=delete-snapshot&parameters={name=snapshot_A}"
200 OK
```

I started a draft branch detailing the schema part enabling discovery of the actions: https://github.com/avade/servicebroker/compare/method-attached-schemas...orange-cloudfoundry:issue114-from-165?expand=1 including the input parameters and the response This is based on avade:method-attached-schemas in #165. The work on detailing the REST endpoints is not started.

Pending issues:
- how long should completed action resources available for query ? should the spec describe this ? Brokers could need to implement a kind of garbage collection on completed actions.
- is there a need/use-cases for update/delete of an action resource ?

**Comment by petereberlein** [comment]
The semantics of deleting an action resource would be something like a cancel - this is already useful for the backup/restore use case.
Updating is less obvious as a typical implementation would not allow updating action parameters during action execution. But who knows what people will use actions for in the future? So we might not specify update for now but reserve it for when a good use case comes up.

**Comment by mattmcneeney** [comment]
There is an ongoing conversation in the [OSBAPI Slack](https://openservicebrokerapi.slack.com/messages/C2S1X842Y) about the state that brokers do/should keep (this has implications on the async bindings and fetching input parameters features). There is support for stateless brokers; it keeps them simple and easy to implement. If brokers were required to keep the results of many actions, perform garbage collection, etc, then this could make them much more complex to develop.

However, with async create, update and delete (*optional*), brokers do already have to maintain some state (remembering the last operation result).

Regarding @gberche-orange's proposal (thanks Guillaume!);
* I like the extensibility of `actions` and the ability to define schemas
* Is there some inconsistency with existing endpoints if the platform is specifying the `action_instances` ID? With the existing async calls, the broker returns an `operation` string to the platform which is used to check the status of the operation via the `last_operation` endpoint.
* To me it makes sense for `PUT` to create a new action, `GET` to fetch the status and `DELETE` to cancel it. I'm not sure we should handle other HTTP verbs for now until we have a strong use case.

Before we dive too deep into the implementation, please could everyone share their current use cases for actions? For CF, the key ones we want to solve asap are backup and restore. What other actions are people finding a need for brokers to support?

**Comment by gberche-orange** [comment]
> Before we dive too deep into the implementation, please could everyone share their current use cases for actions?

The [proposal document "problem" section](https://docs.google.com/document/d/1Dhs4RrvXxyDcmK8LdxbkHcEb_HrpiA4XqVwaIy2sTXg/edit#heading=h.twf85jubj20p) details some use-cases such as:
- stop/start a service instance as to reduce resource consumption without losing data hold by the service. For data services, a stop may trigger an archival of the data to cold storage (such as AWS glacier) and a release of hot/warm computing resources associated to the service instance (storage, RAM, …), a start may fetch data from cold storage and provision computing resources.
- configure the service, such as enable/disable verbose logging.

I refined the problem section with additional details, benefits of actions and non-goals section. Feel free to make suggestions to the google doc or ask for edit permissions.

> I would like to revisit the conversation about supporting B&R as a first class API concept. It is still Pivotals main use case for actions.
> We can still progress with actions as a concept, but I would prioritise solving the B&R case first.

It would be useful to alternative see 1st class support for snapshot/restore in OSB to compare both approaches.

Potential pros for 1st class support for snapshot/restore in OSB w.r.t. broker actions:
- could allow for more friendly user experience (e.g. display a list of available snapshots available to restore from and prompt user to select one) than what simple generic broker actions are likely to enable while keeping complexity in OSB API low.

Potential cons for 1st class support for snapshot/restore in OSB w.r.t. broker actions:
- common B&R denominator among data services is too small, and that a normative B&R be too simplistic to be effective for some services. Would need extensibility to be built-in.

**Comment by nilebox** [comment]
Any updates on this issue?

> I'd like to discuss this at the f2f in a few weeks

I see the following action in the F2F notes: _Define a generic “actions” mechanism - Alex L, Doug_
@avade @duglin do you have any updates or further details on this?

> Before we dive too deep into the implementation, please could everyone share their current use cases for actions?

Our use cases are very similar to the ones already listed: we need a support of _extensible_ set of Operations actions supported by service brokers (backup/restore or anything else service broker specific). So generic mechanism of defining actions would be useful for us.

**Comment by vaikas-google** [comment]
As discussed on the weekly call, here's a draft proposal for making actions be more generic.
https://docs.google.com/document/d/1fVBzqi6OK-lyfjDKADSk35hZD4YUqM2AMtoLOgrDDCE/edit

**Comment by rhodie27** [comment]
As part of my AI from the F2F the attached zip contains an example swagger for generic backup & restore actions, (there are 4 formats, so pick your poison).

It not perfect, but I think it will help to continue the conversation. Things that still need refinement:

1. async representation (easier in openapi 3.0 which isn't fully supported yet, I just added bare bones)
2. security / credential management
3. restore point retention / deletion
4. scheduling (not part of this swagger, but I will upload another with a scheduler action)
5. non-success http status codes
6. better swagger formatting

Next steps:

1. review with osb team
2. incorporate feedback
3. build a sample UI from a OSB API that serves this swagger on request

[bandr-swagger.zip](https://github.com/openservicebrokerapi/servicebroker/files/1295995/bandr-swagger.zip)

**Comment by shpandrak** [comment]
Hi, Regarding the backup and restore actions discussion. While restoring service data to production is an important use case. The more frequent use case is actually restoring service instance backup to a new service instance (lets call it repurpose). In most cases, even when a restore to production is needed, it is common to first repurpose the backup to a new instance in order to perform validation or having both current and restored instances side by side. Current suggestions I have seen treat backups as a sub entity limited to a specific service instance scope. Need to make sure we support the repurpose use case for the API to be useful. 
One option could be to allow creating a blank instance and then running restore but it is problematic since it prevents implementations to do optimizations (e.g. aws rds can restore from snapshot). Seems like backups behave as a top level entity. Anyways, the backup and restore natives/custom actions is a long expected feature - looking forward for seeing it in production 👍

**Comment by nilebox** [comment]
@rhodie27 hi Michael, any updates on the progress of this issue?

**Comment by rhodie27** [comment]
@nilebox, sure. I will have an updated swagger, a server implementation of that swagger and an example UI for actions posted in a couple of days.

**Comment by mogul** [comment]
If anyone's still looking for sample actions outside of B&R, one of the frequent requests we get is "I'd like a way to invalidate the cache please" for https://github.com/18F/cf-cdn-service-broker

**Comment by vaikas-google** [comment]
@rhodie27 have you had a chance to make progress on this?

**Comment by rhodie27** [comment]
waiting on legal @vaikas-google. Let's discuss on the call tomorrow.

**Comment by rhodie27** [comment]
Apologies for the delay, I should not have any more roadblocks going forward.

I put a demo server & ui @ https://github.com/rhodie27/osb-actions-poc to prove out the concept. Please download, build and use to see everything in action and then ask any questions.

In general, everythings looks really good from our point of view, but we can discuss in more detail at the next call and/or F2F.

The next step(s) would be to:

1. Draft the language to get discovery_url in the spec
2. Figure out how/where we want to handle/store the 'suggested' action swaggers, (like backup/restore, schedule, etc.)
3. Draft/finalize said set of suggested actions.

Hopefully we can dedicate a good chunk of time for this at the F2F.

**Comment by nilebox** [comment]
@rhodie27 thanks for the proof of concept, something concrete is much easier to digest.

We are currently working on the support for actions in brokers internally, so this proposal is relevant to us. However, we are focused on supporting custom actions as first-class citizens, supported in the same way as regular backup/restore.

Why do we need completely separate endpoints for each of the predefined actions (backup/restore, start/stop/pause)? Why can't we just make them part of the generic actions API in the similar fashion to `service_instances` and `service_bindings`? i.e. have a `/service_instances/{instance_id}/actions/{action_id}` where you specify the action type in the body similar to the way we currently specify service/plan for instance, for example.

Even if we do want to have distinct sub-URLs for `backups`/`restores`, it's probably good to still have an `/actions/` prefix for all actions just to allow for support of other kinds of features for instances in the future. Actually, we already do have a `last_operation` endpoint (with proposed support for `operations`), i.e. `service_instances/{instance_id}/last_operation`.

Also comments on the spec:
- What about support for asynchronous execution? Are they supported for backups (there is a `GET` endpoint on a single backup)? For restores there is no `GET` endpoint, but they could also take significant time (and we should probably be able to reliably retrieve the status of restore operation even if we lost the connection). Same for start/restart/stop - how do we know what's the current state of the instance and if the operation succeeded?
- REST endpoints are not compliant with the current OSB spec, e.g.:
  - for instance provisioning: `PUT` is for provisioning instance (with client-side uuid provided), `PATCH` for updating it;
  - in actions spec, `POST` is used for creating backups (no client-side uuid provided, server-side id generated instead), `PUT` for updating backups
- What does updating a backup mean? Is it really needed? "Action" sounds like something imperative/immutable which should not support updates (I understand that there's a distinction between purely imperative action like `restart` and `backup` which produces a database dump, but still - trying to be generic)
- Listing backups does sound like something useful, will we be able to support similar semantics for custom actions? 
- `swagger` keyword should be replaced with `open-api` since this is the correct name of the standard (see https://swagger.io/specification/)

> Hopefully we can dedicate a good chunk of time for this at the F2F.

Unfortunately, I won't be able to participate in it in person, but would be happy to connect remotely if you can setup a Zoom/Skype/whatever call.

**Comment by nilebox** [comment]
More comments:
- in addition to above, the problem with `backups` is that it mixes the persistent entity (a database dump) with operation (backup creation/update/delete). It might make sense to have separate `GET` endpoints for them, e.g. `backups/{backup_id}` vs `backups/{backup_id}/last_operation` like we have for instances.
- there is a `resume` action missing (a counterpart for `pause`, similar to `start`/`stop`)

**Comment by rhodie27** [comment]
@nilebox thanks for the feedback, let me see if I can address all of these comments.

> Why do we need completely separate endpoints for each of the predefined actions (backup/restore, start/stop/pause)? Why can't we just make them part of the generic actions API in the similar fashion to service_instances and service_bindings? i.e. have a /service_instances/{instance_id}/service_actions/{action_id} where you specify the action type in the body similar to the way we currently specify service/plan for instance, for example.

1. In the proposal from @vaikas-google, the service provider can really do whatever they want with the returned discovery_url swagger specification. It's owned by them, but we want to provide examples of popular actions. For our examples, we can discuss this, but I think it really depends on the action in question.
2. The backup/restore and stop/start/pause/restart swaggers are just examples to prove out the concept. We need to sit down and really plan them out if we want to own these in addition to the osb spec.

> What about support for asynchronous execution?

Yup, I just didn't add it yet into the example. The question I have for the group is would we want async to look like we have it in the V2 spec, or something like the callbacks feature that swagger pushes.

> REST endpoints are not compliant with the current OSB spec

Understood. I took an older version of an OSB swagger spec just to get this done quicker. I'll update when we get the OSB swagger updated to the current spec version.

> What does updating a backup mean? Is it really needed?

From my pov it is needed if you want to do things like move backup files around, update the expiration times, etc. Is it mandatory? No, but we can talk about this more as we formalize the example.

> Listing backups does sound like something useful, will we be able to support similar semantics for custom actions?

Listing backups is mandatory, (imo), for a backup/restore usecase. Sorry, I don't understand the latter part of your question; can you expand on that?

> swagger keyword should be replaced with open-api

Will fix. Thanks.

> Unfortunately, I won't be able to participate in it in person, but would be happy to connect remotely if you can setup a Zoom/Skype/whatever call.

We will have zoom or something for sure. We can also setup something to talk outside of the F2F.

> in addition to above, the problem with backups is that it mixes the persistent entity (a database dump) with operation (backup creation/update/delete). It might make sense to have separate GET endpoints for them, e.g. backups/{backup_id} vs backups/{backup_id}/last_operation like we have for instances.

Yeah we could probably do that. Imo, the backup resource is the action of requesting a backup, (who, what, when, where, etc). The mechanism to do the actual backup of the service instance is just a property of that request.

> maybe we even should address this issue differently, if we want to support such distinction in a generic way (make imperative actions and persistent entities completely separate), since currently we have both imperative actions (start/stop/restart, restore) and persistent backups at the same level with very different REST semantics.

I think this is what we get with generic actions; the service providers will be able to provide whatever they want in their swaggers. I guess we could try to come up with a standard way to format their apis, in general.

> there is a resume action missing (a counterpart for pause, similar to start/stop)

Yeah, I spent no time on the start/stop/resume/pause stuff... just provide another discovery_url as an example. As part of the next steps above I want to start making these swagger specs more 'official'.

**Comment by rhodie27** [comment]
Also, one additional thing to think about is that the actions URLs do not have to be part of the OSBAPI server; they just happen to be in the example I provided. An example of this could be an Azure service broker sending backup & restore actions to their native api.

**Comment by nilebox** [comment]
> Imo, the backup resource is the action of requesting a backup, (who, what, when, where, etc). The mechanism to do the actual backup of the service instance is just a property of that request.

Agreed, but then if you want to manage database dumps, you need separate actions like `create-backup` and `delete-backup` instead of `PUT` and `DELETE` on a "backup" resource, and have `GET create-backup` and `GET delete-backup`. Then you will be able to tell who and when created and deleted database dump (or even tried to create/delete but failed). With a "backup" resource you don't have such granular history of actions, only the current state of the resource.

> The question I have for the group is would we want async to look like we have it in the V2 spec, or something like the callbacks feature that swagger pushes.

The issue with the current async support in OSB is that there are no mandatory `GET` endpoints (for retrieving results of performed provisioning/binding). That's why in the async binding PR (https://github.com/openservicebrokerapi/servicebroker/pull/334) `GET` endpoints become mandatory.

**Comment by nilebox** [comment]
@rhodie27 @vaikas-google I would like to discuss an alternative proposal for generic actions API we have designed internally.
The most important difference from what was proposed there is that marketplace can support standard and custom operations in exactly the same way, and both can be fully supported by CloudFoundry / Service Catalog without any special code for handling specific types.

The basic idea is that:
- OSB declares generic endpoints for operations (both synchronous and asynchronous) supported by the marketplace, let's call it "conventional API". It also declares common schemas for request/response bits (still allowing for arbitrary input/output payload). This allows the platform to understand the semantics of "action execution", "async action status polling" and so on, in a generic way.
- There is a broker endpoint for declaring a list of supported operations in the OpenAPI (Swagger) format. If the URLs of the operations in the OpenAPI spec are compatible with the "conventional API", and the schemas are based on the common schemas mentioned above, the marketplace provides support for their execution. If not, these types of operations can be still potentially "proxied" by the platform, but without understanding semantics of particular GET/POST/etc request (we can discuss if we actually want to support this extra flexibility).

See draft OpenAPI (Swagger) spec: https://swaggerhub.com/apis/Atlassian/osb-actions/0.1-alpha
And PDF document describing the design in more details with examples
[OSB-Ops.pdf](https://github.com/openservicebrokerapi/servicebroker/files/1474601/OSB-Ops.pdf) (I can transform it into markdown and publish on GitHub, if you like the idea in general).

I can also implement an example of the broker's implementation of this spec with support for backup/restore and start/stop/restart, i.e. the same "popular" operations currently being evaluated, so we can compare them with the proof-of-concept made by @rhodie27.

If there's an interest in this proposal, we can discuss it next week at the SIG call and/or at the F2F.

**Comment by rhodie27** [comment]
A couple of quick comments before I dig in further:

1. If it wasn't clear, the generic actions proposal from @vaikas-google is the google doc above with the changes to the osb api for discovery_url support. Not the swagger I created; that's just an example. The PoC I did was to see if building a dynamic (and performant) UI off of a discovered swagger doc was acceptable. I think that's proved out, so maybe you don't need to do the same here, but others might disagree.
2. Your swagger proposal seems more in line with the "actions a first class citizen" debate; maybe we need to rehash that? The problem I see is that we can't make broker authors follow the external action guidelines; we can only recommend that they do. That's why I figured offering up explicit resources made more sense. If we do this as a first class citizen your way is more appropriate.
3. The biggest thing I'm interested in is standardization of names and properties. I don't want to see one broker implement "backups" another "snapshots" and then another "mysqldumps". I think your swagger still leaves room for that. 
4. Just wondering, how would you handle scheduling jobs in your proposal? E.g. I want to run a backup on a service instance every 6 hours?
5. This looks really nice. Good job.

**Comment by nilebox** [comment]
Thanks for the quick feedback!

> the generic actions proposal from @vaikas-google is the google doc above with the changes to the osb api for discovery_url support

yeah, but my impression was that there is a huge gap between just discovery_url (which essentially lets platform only act as a "proxy" which can only validate the request without understanding semantics and standartization, as far as I understand) and supporting "actions as a first-class citizen". I will probably spend some time on re-reading the google doc and check if I missed something.

> Your swagger proposal seems more in line with the "actions a first class citizen" debate; maybe we need to rehash that? 

Yep, it is. Maybe we do need to rehash that :)

> we can't make broker authors follow the external action guidelines; we can only recommend that they do

Not sure what do you mean by "external guidelines" there, and what's the problem with supporting that. It's possible wrap any specifics into the standardized API, why not? Instance provisioning is also cloud platform specific, yet we were able to define a standardized generic API for that.

> I don't want to see one broker implement "backups" another "snapshots" and then another "mysqldumps". I think your swagger still leaves room for that.

Yes, you're right, this concern is not addressed in my swagger. Maybe just written guidelines with convention on the naming in existing brokers is enough? Also, I'm not sure if you can standardize properties. SQS doesn't support backups at all, and some other resources could have more specific terms like "snapshots", "replication" or whatever else rather than just "backups". Even MySQL could support "native dumps" as well as "SQL logs" or even export to XML. I think this flexibility is good. OSB services are not just merely different types of databases, they can be anything.
But I agree that it's nice to call same things with the same names, no doubt.

> Just wondering, how would you handle scheduling jobs in your proposal? E.g. I want to run a backup on a service instance every 6 hours?

That's an interesting use case, I haven't thought about it. Several options come to my mind:
- "Do nothing": keep scheduling out of scope of the broker - leave it to external service to invoke platform/broker when the backup should be performed
- Define the `schedule_backup` action. The only problem with this approach is that it's not trivial how we'll check the history of all performed backups or a list of all available database dumps (backup results). One of the options is for `schedule_backup` to return a custom URL in the response, with the web UI or REST endpoint where user can go and fetch this information.

**Comment by ash2k** [comment]
> Define the schedule_backup action. The only problem with this approach is that it's not trivial how we'll check the history of all performed backups or a list of all available database dumps (backup results).

It could be an action `list_backups` that returns a list of all existing backups. Whether they were created via an OSB API call or not does not matter.

**Comment by nilebox** [comment]
@rhodie27 sorry for the noise, I feel dumb now since I totally misunderstood the purpose of your proof-of-concept work and completely missed the auto-generated UI part. I was confused by very different operation implementations (backup as a complex resource, restart as a simple imperative action), and had an impression that the goal is to standardize API for popular actions.

Now I understand that the purpose of this exercise is an evaluation of building UI based on arbitrary swagger spec supplied by the broker. I do think that async operation handling, as well as handling errors is a must for evaluation though.

This approach is very similar to what we've tried to achieve with our proposal. Basically, our proposal is an extension of yours, it also requires swagger spec for supported operations, with an extra (opt-out) standardization for supporting async jobs, status response and error outputs which will help us with the tooling to support these features. So yes, you're absolutely right that this proposal brings back the "actions a first class citizen" debate, but with the flexibility and opt-out.

With such standardization, CF or Service Catalog (or a separate service) will be able to not only initiate the asynchronous action (and then force user to use a separate action for manual polling), but will also be able to poll for status until the operation succeeds / fails, and display messages and errors with a nice UI or logging messages / output.

By "opt-out" I mean that if a particular action is hard to wrap into a standardized API (as you mentioned above), it can be exposed directly with an arbitrary contract. This will mean that we will need to utilize your UI for supporting such flexibility, but with less help than with the standardized API. Polling for asynchronous action can be a pain for example, with no indication of the progress in the UI, no smart error formatting and so on.

> The biggest thing I'm interested in is standardization of names and properties.

I still fail to see how do you achieve that with just `discovery_url` approach though.

**Comment by nilebox** [comment]
> The PoC I did was to see if building a dynamic (and performant) UI off of a discovered swagger doc was acceptable. I think that's proved out

Not sure if I follow, so I would like to ask for some clarification there.
For completely dynamic UI there should be nothing about specific actions (backup, restore etc) in the UI, right? It should be only based on the supplied swagger spec and nothing else. However in the [osb.js](https://github.com/rhodie27/osb-actions-poc/blob/master/ui/js/osb.js) file I see that there is a switch-case over hardcoded supported actions, and even specific methods for modifying backups ([osb.js#L274](https://github.com/rhodie27/osb-actions-poc/blob/c8a8106094a6b5492a8a6425fd16b1ef73fc0eb9/ui/js/osb.js#L274)). So if I supply some random swagger, the UI won't work I suppose. Was it just a temporary quick hack, and could it be rewritten to a completely generic code?

**Comment by vaikas-google** [comment]
as far as naming mismatches, that's the point of adheres_to. There should be an agreed upon format for these, but imho they don't belong to OSB, they could be defined as part of another set of people.

**Comment by rhodie27** [comment]
> Was it just a temporary quick hack, and could it be rewritten to a completely generic code?

Yes, quick hack. It should work with any similar swagger files, regardless of the backup/restore function names.

And yes, w/ regards to names & properties I was hoping something could be defined outside of the spec, but maybe in the same org. Either way, it would needed to be in, IMO, with both of these specs. And on that note, is this something we should discuss in detail on the call on Tuesday or should we setup a separate / more specific call before the F2F?

**Comment by vaikas-google** [comment]
Yeah, in my head and I thought we talked about this in the last f2f, the
adheres_to would be a pointer to a spec defining the semantics of the
supported actions. There's no need to have one, but defining use case
specific specs (for example backup) seems like it's best left to folks who
are subject matter experts and the OSB spec should make it possible to
support various use cases but not be involved in defining those semantics.
By utilizing adheres_to there should be no issues with the naming conflicts
or differing interpretations of what actions mean. If I understand it
correctly, it seems like the one issue being described here is the calling
semantics of actions and in Nail proposal this is the generic invocation
mechanism for asynchronous jobs and being able to poll them for status as
well as canceling. We chatted about this a bit yesterday offline and I do
think it would be beneficial for us to agree on the goals and design
parameters before heading down the path too far.
I had made a design decision that I think might be at odds here:
 * Broker should not be necessarily the one implementing the actions. Hence
there's no additional endpoints required on the broker. Thinking here was
that a broker could implement management functions but would allow for
offloading these

**Comment by vaikas-google** [comment]
Yeah, in my head and I thought we talked about this in the last f2f, the
adheres_to would be a pointer to a spec defining the semantics of the
supported actions. There's no need to have one, but defining use case
specific specs (for example backup) seems like it's best left to folks who
are subject matter experts and the OSB spec should make it possible to
support various use cases but not be involved in defining those semantics.
By utilizing adheres_to there should be no issues with the naming conflicts
or differing interpretations of what actions mean. If I understand it
correctly, it seems like the one issue being described here is the calling
semantics of actions and in Nail proposal this is the generic invocation
mechanism for asynchronous jobs and being able to poll them for status as
well as canceling. We chatted about this a bit yesterday offline and I do
think it would be beneficial for us to agree on the goals and design
parameters before heading down the path too far.

**Comment by rhodie27** [comment]
Let's deep-dive into the topic on Tuesdays call and if we need more conversation maybe we can do it prior to the F2F.

**Comment by vaikas-google** [comment]
SGTM.

**Comment by shashankmjain** [comment]
I feel, the OSB Actions API should also allow creation of constructs needed for services like (LoadBalancer Endpoints, Security Groups,Network Policies). Currently they are achieved as part of the service provisioning calls, but I feel the concepts like Security Groups, Service endpoints should be decoupled from the service provisioning itself.

**Comment by rhodie27** [comment]
Based on the F2F feedback I put together a list of action use cases and how they would be implemented using this proposal. That can be found @ https://docs.google.com/document/d/1o-JuW0kVH6TIk8i-1nXas6767xyfuUti4p7St-BC-aM/edit?usp=sharing

I think the next steps are:

1. Bring everyone interested up to speed on the use cases, (hopefully including UI people)
2. Agree on terminology, (adheres_to, discovery_url, management_api)
3. Determine if adheres_to is necessary in the first iteration of this.
4. Look at bulk operations, (maybe move mgmt_api to plan).
5. Can this be used for service actions? (https://github.com/openservicebrokerapi/servicebroker/issues/376)
6. Compile a full list of potential actions, (if desired)
7. Start working on the spec language.

What I would like to do is host meetings outside of the weekly call, (with anyone interested), to start driving this to completion. Let's discuss who & the frequency on today's call.

**Comment by rhodie27** [comment]
Notes from the sync up on actions on 12/14/17:

**1.	Bring everyone interested up to speed on the use cases, (hopefully including UI people)**

- @nilebox is going to add his use case of a generic api for many actions into the use case format. 
  -  He thinks it'll fit fine and is happy with the general direction.
-  Jessica gave some insight into the UI flow. She will get more time after the holidays to review.
- We all agreed we need more UI folks to verify the flow.

**2.	Agree on terminology, (adheres_to, discovery_url, management_api)**

- Current consensus:
  - _extention_api_ instead of management_api
  - _discovery_url_ stays as is
  - _adheres_to_ stays as is
  - @vaikas-google ESL is solid

**3.	Determine if adheres_to is necessary in the first iteration of this.**
- We agreed that _adheres_to_ should be necessary in the first iteration of actions.
- Agreed that _adheres_to_ should also be mandatory; if you don't follow a spec you can use a junk URI.
- Do not want to leave it out and then add it in later.
	
**4.	Look at bulk operations, (maybe move mgmt_api to plan).**
- Started the conversation, but we went back to the UI flow and if it worked in general.

**Other Stuff

- @duglin started an outline of the intended schema:
  - https://docs.google.com/document/d/1o-JuW0kVH6TIk8i-1nXas6767xyfuUti4p7St-BC-aM/edit?ts=5a319720#heading=h.1ql3y8ouad77

- @rhodie27 to schedule another call for next week, (at a proper time for the London folks).

**Comment by ash2k** [comment]
👍
And thanks for the notes!

**Comment by vaikas-google** [comment]
For #3, middle point:

   - Agreed that *adheres_to* should also be mandatory; if you don't follow
   a spec you can use a junk URI.

You could also probably just point it to discovery_url :)

**Comment by nilebox** [comment]
Just published an example of our use case with OpenAPI spec for generic operations with semi-strict contract and async execution support: https://docs.google.com/document/d/1o-JuW0kVH6TIk8i-1nXas6767xyfuUti4p7St-BC-aM/edit?ts=5a31a014#heading=h.kss2koahby2k

Seems to work for us 👍

**Comment by fmui** [comment]
Could we allow the `discovery_url` also to be relative to the broker URL? It should be easy for the cloud controller to compile the URL, but it may be difficult/cumbersome for the broker (for example if it resides behind a reverse proxy).

**Comment by duglin** [comment]
If we make it a URI-reference per the URI spec then I think that should cover it.

**Comment by fmui** [comment]
Can we then rename `discovery_url` to `discovery_uri`?

**Comment by n3wscott** [comment]
Do you want URI because you want to allow data: encoded document or other
such thing? It looks like swagger / Open API only allow for URL:
https://swagger.io/docs/specification/api-host-and-base-path/

**Comment by fmui** [comment]
> Do you want URI because you want to allow data: encoded document or other
> such thing? It looks like swagger / Open API only allow for URL:
> https://swagger.io/docs/specification/api-host-and-base-path/

The strange thing is that Open API talks about URLs but refers to the URI spec (RFC 3986).
Open API allows relative paths, which are not covered in the URL spec.

**Comment by duglin** [comment]
yup I noticed the same thing for some other work recently, so I'm opt for URI toos

**Comment by rhodie27** [comment]
Notes from the 1/8 sync up on Actions, (apologies for the delay):

**1.  Continued the debate on adheres_to:** 
- should it be mandatory and point to discovery_url or just be optional.
- will people use it to push standardized extentions to the api.
- should we just leave it out of the first release of actions.

**2. We should probably start calling this feature extensions instead of actions**
- this mechanism is a way to build in actions, but also will support general extensions to api.

**3. UI developer feedback**
- Jessica from Pivotal really wants a list of the action names prior to getting the swagger doc.
- Building a pregenerated UI from adheres_to is not something she is interested in.

**4. General feedback**
- We should just move this to validating by implementation; we won't know the major issues until we get some feedback

**5. Did not get to the bulk operation conversation.** 

**6. Need to have a conversation on where the actions should be returned** 
- plan vs service instance
- E.g. do we really want to say that actions can differ from service instance to service instance?

**7. Decided we would do one more call to discuss bulk operations and plan vs service instance and then get this into the spec behind a feature gate.**
- Mike is going to set this up for Wed 1/17 at 12:30pm EST
- We are going to start getting a PR together for this regardless of the outcome of that meeting.

Recording (104 MB): 
https://zoom.us/recording/play/Ogc_DAXBOiBt1bce7Djyv65ZTw2KVRDU-aNSz-AjFx2QcYOsad3zzwKt7OmcH6hy

Audio Only-1 (28 MB): 
https://zoom.us/recording/play/NTQFO3-JdpXKqU4xs7R7JX8q2XcLcedUrY3WJsGpLHlL8DSMyHtIJDE2sDDnzAX9

**Comment by cforce** [comment]
> 6. Need to have a conversation on where the actions should be returned
> E.g. do we really want to say that actions can differ from service instance to service instance?
If the actions supported by the implementing service party are first queried with instantiation, you can't make it part of the "static" plan, ca you? I mean if the cloud services offers additional functions over the time. If an intance is created, it's based on a version of the instance at creation time e.g. v1. A second instance one week later miggt be instantiated on v2, while first instance is still not upgraded. If instance of v2 has more actions a than sinatnce of v1  the still have the same plan, doesn't they? Plan's does not represent the version of functional service actions which is available with a current instance of a service plan, odes it?

**Comment by rhodie27** [comment]
You wouldn't need to add a new plan to add new actions. You simply add/update the discovery_url or the openapi spec that it points to.

Your "functions over time" use case is valid and interesting.

Regardless, we're going to move forward with actions at the service instance for now.

**Comment by rhodie27** [comment]
Notes from the 1/17 sync up on Extensions, :

**1.  Bulk Operations** 
- this might be supported today, but the work is on the platform side to support.
- we won't know if there is a use case for this until we move to validation
- harder to take things away (from the spec) than to add them, e.g. let's not do too much at the start.

**2. Service Plan vs Service Instance**
- state of a service instance might change the instance, e.g. you shouldn't be able to pause a paused service instance.
- using plan make it harder & more unreliable to handle service instance parameters

**3. Ville & Mike sync up on PR under the following terms**
- Include adheres_to
- Use term 'extentions'
- discovery_url returned via service instance
- need to clearly state what credentials should be used to initiate the actions, e.g. platform or other

**4. General feedback**
- We should just move this to validating by implementation; we won't know the major issues until we get some feedback

**5. Jeremy @ MSFT** 
- first usecase would be Backup Operations 
- SQL DB - provide a pass through to the native Backup/Restore api
- next would be others around Day 2 Operations
- would probably use this proposal right away

**6. John @ Redhat / Ansible** 
- Using actions to execute an ansible task
- Backup / day 2 /  is a priority
- Would be able to use in March / contingent on service catalog.

**7. Retrospective**
- Good use of outside sessions and a way to engage community
- Got good feedback / participation from people that don't attend the weekly status call
- We should use this is a vehicle going forward

Recording:
Mike forgot to hit record.

**Comment by avade** [comment]
Thanks for the detailed notes @rhodie27. Looking forward to reading the PR.

> discovery_url returned via service instance

Glad to hear this was the outcome, I also believe this is the right level for actions to be exposed functionally.

I wonder if a plan could advertise the actions in a way that could be consumed programmatically in a catalog. To start with the description/metadata field could work, at least some kind of community standard.

**Comment by mattmcneeney** [comment]
Renaming this to match PR #431

**Comment by Samze** [comment]
After a period of inactivity we are now revisiting generic instance extensions. There is a new [proposal doc here](https://docs.google.com/document/d/1OKlsQKUab3yvOyQjvh8HFFcIfkILyFMaq0xngTdTcek/edit).

**Comment by mattmcneeney** [comment]
I can't assign @Samze to this, but this is assigned to him :)

**Comment by Samze** [comment]
Replaced by #670
